include/lbann/data_readers/data_reader_jag_conduit.hpp

```diff
@@ -216,19 +216,19 @@ class data_reader_jag_conduit : public generic_data_reader {
   std::string get_description() const;
 
   /// Return the image simulation output of the i-th sample
-  std::vector<cv::Mat> get_cv_images(const size_t i, conduit::Node& sample) const;
+  std::vector<cv::Mat> get_cv_images(const size_t i, conduit::Node& sample);
```

Comment:
I know modern compilers *may* get around this, but returning a local object can result in a copy, which is not efficient. I'd like to see some discussion on this -- perhapsI'm behind the times c++ wise.

Reply:
A compiler _might_ be able to elide the copy here (and C++17 has specific requirements about copy elision being required, not as an optimization, but as part of the language semantics to avoid intermediate materialization of values). I'm not certain because the implementation of these methods is complicated enough, I'm not sure that the elision requirements would apply. @benson31 probably has more insight here.

Reply:
It will never be a copy. In the best case, the copy is elided (I'd actually have to read the function to see if it fits the requirements for the C++17 guarantee). In the worst case, the vector is _moved_; in C++11+ it should never be copied. This "worst case" means that 3-4 pointer-sized things are copied (24-32 bytes), depending on the STL impl being used. See also my comments on #888.